### ISSUE TITLE
Serious Flaw in the New Frost Theme

### ISSUE THREAD

**Issue by tsujan** [lxde/lxqt#1052]
The older version of Frost was OK. In the new version, it seems that the designer's intention was to invert colors by assigning `palette(text)` to backgrounds and `palette(window)` to texts. *That's wrong!* It may work with simple themes and simple style engines but once an SVG-based style engine is used, the problem will be seen.

Since most themes and style engines are simple (in a sense), you might say that the problem is in other ones. But that would be simply incorrect. In Qt, each color palette has its specific purpose, on which all GUI apps rely. You couldn't just use them freely and expect the result to look OK. Please see http://doc.qt.io/qt-4.8/qpalette.html.

The older Frost was (mostly) OK because it assigned specific colors correctly and in those places that it used palettes, it did so right.

In short, color inversion was a wrong idea. Please revert to the older version or modify it appropriately!

**Comment by tsujan** [comment]
Moreover, if you use `palette(highlight)` for the background of a widget, please also use `palette(highlighted-text)` for its text! (→ http://doc.qt.io/qt-4.8/qpalette.html)

**Comment by intialonso** [comment]
Yeah. The intention was to make the theme adapt to the widgets one (and avoid hardcoded colors), while keeping the panel dark (thats the reason for the invertion).

You find this problem just in the SVG engine? Can you provide screenshots?

**Comment by tsujan** [comment]
Right now I'm programming under KDE and can't make a screenshot.

However, it's not necessarily limited to SVG-based style engines (white text on white bg for dark themes) -- those engines may just easily reveal such issues. I'm pretty sure one can make a valid QtCurve theme, that creates dark text on dark bg with the new Frost. Once a rule of color palettes is broken, it'll be hard to know what will happen ;)

Have you ever seen color inversion with style sheets?

BTW, it's really not the designer's fault. I might try to do the same too ;) Nowhere in Qt doc the intricacy of QStyle is explained, perhaps except for the QStyle Class doc itself.

**Comment by intialonso** [comment]
Do you have any suggestion as how to achieve this result without broke anything?

I thing (visually) that hard coded color in our main theme will lead to inconsistencies with the global system look (depending on the user). I mean, the original background, highlight and font colors work ok with the Numix frost theme, but not so nice with anything else, it will look alien when paired with other widgets styles.

Im not sure, I will like to keep the theme with some adaptive qualities, but Im worried about your concerns.

**Comment by tsujan** [comment]
AFAIK, it's simply impossible to switch window and text color palettes and have a stylesheet that works OK everywhere.

If you want the panel to match the global theme, the only way is setting the same color palettes in the stylesheet but then, why setting them at all?

Take a look at KDE: there are two themes, i.e. the widget theme (or color scheme, as they call it in KDE) and the Plasma theme. It's up to the user to match them or to use a dark Plasma theme with a light widget style, or conversely. When you change the color scheme, it doesn't have any effect on the Plasma, and conversely.

The older Frost was OK: a dark scheme was used, as in gnome-shell or like the dark Breeze Plasma theme. If a user wanted a light theme, he/she could apply another LXQt theme.

**Comment by tsujan** [comment]
In addition to what I said above, in the new Frost, you have:
````
#MainMenu QLineEdit {
    border: none;
    margin: 10px;
    padding: 3px;
    background: palette(highlighted-text);
    color: palette(text);
}
````
And, SVG-based or not, there's no guarantee the `palette(highlighted-text)` and `palette(text)` have contrast with each other. If the backround is changed to `palette(window)`, they'll always have enough contrast. You could easily make a QtCurve theme that shows my point.

Anyway, I would avoid color inversion for the above-mentioned reasons. But if you still insist on using it, please search for similar cases and correct them.

**Comment by tsujan** [comment]
I took a look at `lxqt-panel.qss` and corrected some contrast issues I found: https://dl.dropboxusercontent.com/u/7793460/lxqt-panel.qss

It should work OK mostly (which is enough -- obsession isn't fun) but please check it for errors because I edited it hastily.

**Comment by stefonarch** [comment]
Looks quite good, search field has just no evidence, all black. There is one thing to fix too, titles are invisibile in popup menu as below second shot:
![schermata01](https://cloud.githubusercontent.com/assets/10681413/15270350/cb85e83e-1a1c-11e6-815b-5c9150524f43.png)
![schermata](https://cloud.githubusercontent.com/assets/10681413/15270351/cb898070-1a1c-11e6-9732-92b6e77b8619.png)

**Comment by tsujan** [comment]
@stefonarch Thanks for testing!

> ... titles are invisibile in popup menu

I edited it in a hurry. Tomorrow I'll check it again, or @intialonso might correct it before I do.

**Comment by intialonso** [comment]
Go ahead @tsujan Im a little overloaded with my real job this month :(

**Comment by tsujan** [comment]
@intialonso I corrected the issue that @stefonarch mentioned. It should be OK now. With https://dl.dropboxusercontent.com/u/7793460/lxqt-panel.qss, your idea of color inversion can be rescued to a great extent.

There still remain three issues that can 't be avoided with color inversion:

(1) Most light themes use pitch-black as their text color, so the panel bg will be pitch-black with them. This may not be a seen as a problem at all.

(2) The user may like to use dark icons with a global light theme or light icons with a dark theme. With color inversion, menu icons would be terrible ;) The solution would be using another LXQt theme, so it wouldn't be a big deal, IMO.

(3) It's still possible that some themes show minor problems but, IMO, they would be negligible. Most texts should be readable with all themes now.

**Comment by stefonarch** [comment]
Titles in panel pop-up menu are fixed for 5 style-widgets now, thanks.

**Comment by Houkime** [comment]
Well, is there a version of Frost with fixed black colors?
Not Numix frost though - don't need gray at the bottom.

Invertion kinda ruins all recolorable icons and stuff (they are black for white widgets and vice versa and hence are guaranteed to be invisible on the panel)

**Comment by Houkime** [comment]
Well, if someone needs one with fixed black-n-white I made one for myself.
just rename .txt to .qss and replace original file.
I use it with darkened fusion widgets (via qt5-ct). Looks ok.
[lxqt-panel.txt](https://github.com/lxde/lxqt/files/1806650/lxqt-panel.txt)
